idaes/apps/caprese/data_manager.py

```
@@ -0,0 +1,361 @@
+        # to something more meaningful, such as temperature_actual.
+        rename_map = ComponentMap()
+    var_dict = OrderedDict()
+    hash_set = set()
+    var_dict["iteration"] = []
+    for var in variables:
+        # TODO: In this case, do we need to make sure we have the
+        # same indexing sets? 
+        # KHL:I think we can leave the check alone here and do it when adding values. 
+        #
+        #idx_set = var.index_set()
+        #set_hash = hash(tuple(idx_set))
+        #if len(hash_set) == 0:
+        #    hash_set.add(set_hash)
+        #elif set_hash not in hash_set:
+        #    raise ValueError()
+        if var in rename_map:
+            key = rename_map[var]
+        else:
+            if var.is_reference():
+                # NOTE: This will fail if var is a reference-to-list
+                # or reference-to-dict.
+                key = str(ComponentUID(var.referent))
+            else:
+                key = str(ComponentUID(var))
+        var_dict[key] = []
+        dataframe = pd.DataFrame(var_dict)
+        # Change the type of iteration column from float to integer
+        dataframe["iteration"] = dataframe["iteration"].astype(int)
+    return dataframe
+
+
+def add_variable_values_to_dataframe(
+        dataframe,
+        variables,
+        iteration,
+        time_subset=None,
+        rename_map=None,
+        initial_time=None,
+        time_map=None,
+        ):
+    '''
+    This function appends data to the dataframe. 
+
+    Parameters
+    ----------
+    dataframe : pandas dataframe that we want to append the data to.
+    variables : list of variables of interest.
+    iteration : current iteration.
+    time_subset : time indices of interset in the variables.
+    rename_map : dictionary or componentmap that maps the variable to a 
+                    specific name string.
+    initial_time : initial time for this data saving process.
+    time_map : map the time in time_subset to real time points
+    '''
```

I like this function. However, it can be dangerous if `time_subset` contains zero, and the previous dataframe is non-empty. Should we raise an error in this case?

---

Yes, we should, but I think the error should be raised when the `initial_time` and `time_map` are also `None` at the same time.

---

Good point!